File: 2764-maximum-number-of-fish-in-a-grid/2764-maximum-number-of-fish-in-a-grid.js
/**
 * @param {number[][]} grid
 * @return {number}
 */

// DFS & BFS, TC: O(M x N), SP: O(M x N)
var findMaxFish = function (grid) {
    const m = grid.length;
    const n = grid[0].length;
    const directions = [[0, 1], [1, 0], [0, -1], [-1, 0]];
    let result = 0;

    const visited = Array.from({ length: m }, () => Array(n).fill(false));
    //console.log(visited);

    const isValid = (x, y) => x >= 0 && y >= 0 && x < m && y < n && grid[x][y] > 0 && !visited[x][y];

    // DFS implementation
    const dfs = (x, y) => {
        visited[x][y] = true;
        let fish = grid[x][y];

        for (const [dx, dy] of directions) {
            const nx = x + dx;
            const ny = y + dy;

            if (isValid(nx, ny)) {
                fish += dfs(nx, ny);
                console.log(fish);
            }
        }

        return fish;
    };

    // BFS implementation
    const bfs = (x, y) => {
        const queue = [[x, y]];
        visited[x][y] = true;
        let fish = 0;

        while (queue.length > 0) {
            const [cx, cy] = queue.shift();
            fish += grid[cx][cy];

            for (const [dx, dy] of directions) {
                const nx = cx + dx, ny = cy + dy;
                if (isValid(nx, ny)) {
                    visited[nx][ny] = true;
                    queue.push([nx, ny]);
                }
            }
        }
        console.log(fish);
        return fish;
    };

    let maxFish = 0;
    for (let i = 0; i < m; i++) {
        for (let j = 0; j < n; j++) {
            if (grid[i][j] > 0 && !visited[i][j]) {
                maxFish = Math.max(maxFish, bfs(i, j));
            }
        }
    }

    //console.log(visited);

    return maxFish;
};

File: 2764-maximum-number-of-fish-in-a-grid/2764-maximum-number-of-fish-in-a-grid.py
class Solution:
    def findMaxFish(self, grid: List[List[int]]) -> int:
        m, n = len(grid), len(grid[0])
        directions = [(0, 1), (1, 0), (0, -1), (-1, 0)]
        visited = [[False] * n for _ in range(m)]

        def isValid(x, y):
            return (
                0 <= x < m
                and 0 <= y < n
                and grid[x][y] > 0
                and not visited[x][y]
            )

        def dfs(x, y):
            visited[x][y] = True
            fish = grid[x][y]

            for dx, dy in directions:
                nx, ny = x + dx, y + dy
                if isValid(nx, ny):
                    fish += dfs(nx, ny)

            return fish

        maxFish = 0
        for i in range(m):
            for j in range(n):
                if grid[i][j] > 0 and not visited[i][j]:
                    maxFish = max(maxFish, dfs(i, j))

        return maxFish
